perl5/man/man3/re.3
.\" Automatically generated by Pod::Man 4.14 (Pod::Simple 3.43)
.\"
.\" Standard preamble:
.\" ========================================================================
.de Sp \" Vertical space (when we can't use .PP)
.if t .sp .5v
.if n .sp
..
.de Vb \" Begin verbatim text
.ft CW
.nf
.ne \\$1
..
.de Ve \" End verbatim text
.ft R
.fi
..
.\" Set up some character translations and predefined strings.  \*(-- will
.\" give an unbreakable dash, \*(PI will give pi, \*(L" will give a left
.\" double quote, and \*(R" will give a right double quote.  \*(C+ will
.\" give a nicer C++.  Capital omega is used to do unbreakable dashes and
.\" therefore won't be available.  \*(C` and \*(C' expand to `' in nroff,
.\" nothing in troff, for use with C<>.
.tr \(*W-
.ds C+ C\v'-.1v'\h'-1p'\s-2+\h'-1p'+\s0\v'.1v'\h'-1p'
.ie n \{\
.    ds -- \(*W-
.    ds PI pi
.    if (\n(.H=4u)&(1m=24u) .ds -- \(*W\h'-12u'\(*W\h'-12u'-\" diablo 10 pitch
.    if (\n(.H=4u)&(1m=20u) .ds -- \(*W\h'-12u'\(*W\h'-8u'-\"  diablo 12 pitch
.    ds L" ""
.    ds R" ""
.    ds C` ""
.    ds C' ""
'br\}
.el\{\
.    ds -- \|\(em\|
.    ds PI \(*p
.    ds L" ``
.    ds R" ''
.    ds C`
.    ds C'
'br\}
.\"
.\" Escape single quotes in literal strings from groff's Unicode transform.
.ie \n(.g .ds Aq \(aq
.el       .ds Aq '
.\"
.\" If the F register is >0, we'll generate index entries on stderr for
.\" titles (.TH), headers (.SH), subsections (.SS), items (.Ip), and index
.\" entries marked with X<> in POD.  Of course, you'll have to process the
.\" output yourself in some meaningful fashion.
.\"
.\" Avoid warning from groff about undefined register 'F'.
.de IX
..
.nr rF 0
.if \n(.g .if rF .nr rF 1
.if (\n(rF:(\n(.g==0)) \{\
.    if \nF \{\
.        de IX
.        tm Index:\\$1\t\\n%\t"\\$2"
..
.        if !\nF==2 \{\
.            nr % 0
.            nr F 2
.        \}
.    \}
.\}
.rr rF
.\"
.\" Accent mark definitions (@(#)ms.acc 1.5 88/02/08 SMI; from UCB 4.2).
.\" Fear.  Run.  Save yourself.  No user-serviceable parts.
.    \" fudge factors for nroff and troff
.if n \{\
.    ds #H 0
.    ds #V .8m
.    ds #F .3m
.    ds #[ \f1
.    ds #] \fP
.\}
.if t \{\
.    ds #H ((1u-(\\\\n(.fu%2u))*.13m)
.    ds #V .6m
.    ds #F 0
.    ds #[ \&
.    ds #] \&
.\}
.    \" simple accents for nroff and troff
.if n \{\
.    ds ' \&
.    ds ` \&
.    ds ^ \&
.    ds , \&
.    ds ~ ~
.    ds /
.\}
.if t \{\
.    ds ' \\k:\h'-(\\n(.wu*8/10-\*(#H)'\'\h"|\\n:u"
.    ds ` \\k:\h'-(\\n(.wu*8/10-\*(#H)'\`\h'|\\n:u'
.    ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'^\h'|\\n:u'
.    ds , \\k:\h'-(\\n(.wu*8/10)',\h'|\\n:u'
.    ds ~ \\k:\h'-(\\n(.wu-\*(#H-.1m)'~\h'|\\n:u'
.    ds / \\k:\h'-(\\n(.wu*8/10-\*(#H)'\z\(sl\h'|\\n:u'
.\}
.    \" troff and (daisy-wheel) nroff accents
.ds : \\k:\h'-(\\n(.wu*8/10-\*(#H+.1m+\*(#F)'\v'-\*(#V'\z.\h'.2m+\*(#F'.\h'|\\n:u'\v'\*(#V'
.ds 8 \h'\*(#H'\(*b\h'-\*(#H'
.ds o \\k:\h'-(\\n(.wu+\w'\(de'u-\*(#H)/2u'\v'-.3n'\*(#[\z\(de\v'.3n'\h'|\\n:u'\*(#]
.ds d- \h'\*(#H'\(pd\h'-\w'~'u'\v'-.25m'\f2\(hy\fP\v'.25m'\h'-\*(#H'
.ds D- D\\k:\h'-\w'D'u'\v'-.11m'\z\(hy\v'.11m'\h'|\\n:u'
.ds th \*(#[\v'.3m'\s+1I\s-1\v'-.3m'\h'-(\w'I'u*2/3)'\s-1o\s+1\*(#]
.ds Th \*(#[\s+2I\s-2\h'-\w'I'u*3/5'\v'-.3m'o\v'.3m'\*(#]
.ds ae a\h'-(\w'a'u*4/10)'e
.ds Ae A\h'-(\w'A'u*4/10)'E
.    \" corrections for vroff
.if v .ds ~ \\k:\h'-(\\n(.wu*9/10-\*(#H)'\s-2\u~\d\s+2\h'|\\n:u'
.if v .ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'\v'-.4m'^\v'.4m'\h'|\\n:u'
.    \" for low resolution devices (crt and lpr)
.if \n(.H>23 .if \n(.V>19 \
\{\
.    ds : e
.    ds 8 ss
.    ds o a
.    ds d- d\h'-1'\(ga
.    ds D- D\h'-1'\(hy
.    ds th \o'bp'
.    ds Th \o'LP'
.    ds ae ae
.    ds Ae AE
.\}
.rm #[ #] #H #V #F C
.\" ========================================================================
.\"
.IX Title "re 3"
.TH re 3 "2022-04-17" "perl v5.36.0" "Perl Programmers Reference Guide"
.\" For nroff, turn off justification.  Always turn off hyphenation; it makes
.\" way too many mistakes in technical documents.
.if n .ad l
.nh
.SH "NAME"
re \- Perl pragma to alter regular expression behaviour
.SH "SYNOPSIS"
.IX Header "SYNOPSIS"
.Vb 2
\&    use re \*(Aqtaint\*(Aq;
\&    ($x) = ($^X =~ /^(.*)$/s);     # $x is tainted here
\&
\&    $pat = \*(Aq(?{ $foo = 1 })\*(Aq;
\&    use re \*(Aqeval\*(Aq;
\&    /foo${pat}bar/;                # won\*(Aqt fail (when not under \-T
\&                                   # switch)
\&
\&    {
\&        no re \*(Aqtaint\*(Aq;             # the default
\&        ($x) = ($^X =~ /^(.*)$/s); # $x is not tainted here
\&
\&        no re \*(Aqeval\*(Aq;              # the default
\&        /foo${pat}bar/;            # disallowed (with or without \-T
\&                                   # switch)
\&    }
\&
\&    use re \*(Aqstrict\*(Aq;               # Raise warnings for more conditions
\&
\&    use re \*(Aq/ix\*(Aq;
\&    "FOO" =~ / foo /; # /ix implied
\&    no re \*(Aq/x\*(Aq;
\&    "FOO" =~ /foo/; # just /i implied
\&
\&    use re \*(Aqdebug\*(Aq;                # output debugging info during
\&    /^(.*)$/s;                     # compile and run time
\&
\&
\&    use re \*(Aqdebugcolor\*(Aq;           # same as \*(Aqdebug\*(Aq, but with colored
\&                                   # output
\&    ...
\&
\&    use re qw(Debug All);          # Same as "use re \*(Aqdebug\*(Aq", but you
\&                                   # can use "Debug" with things other
\&                                   # than \*(AqAll\*(Aq
\&    use re qw(Debug More);         # \*(AqAll\*(Aq plus output more details
\&    no re qw(Debug ALL);           # Turn on (almost) all re debugging
\&                                   # in this scope
\&
\&    use re qw(is_regexp regexp_pattern); # import utility functions
\&    my ($pat,$mods)=regexp_pattern(qr/foo/i);
\&    if (is_regexp($obj)) {
\&        print "Got regexp: ",
\&            scalar regexp_pattern($obj); # just as perl would stringify
\&    }                                    # it but no hassle with blessed
\&                                         # re\*(Aqs.
.Ve
.PP
(We use $^X in these examples because it's tainted by default.)
.SH "DESCRIPTION"
.IX Header "DESCRIPTION"
.SS "'taint' mode"
.IX Subsection "'taint' mode"
When \f(CW\*(C`use re \*(Aqtaint\*(Aq\*(C'\fR is in effect, and a tainted string is the target
of a regexp, the regexp memories (or values returned by the m// operator
in list context) are tainted.  This feature is useful when regexp operations
on tainted data aren't meant to extract safe substrings, but to perform
other transformations.
.SS "'eval' mode"
.IX Subsection "'eval' mode"
When \f(CW\*(C`use re \*(Aqeval\*(Aq\*(C'\fR is in effect, a regexp is allowed to contain
\&\f(CW\*(C`(?{ ... })\*(C'\fR zero-width assertions and \f(CW\*(C`(??{ ... })\*(C'\fR postponed
subexpressions that are derived from variable interpolation, rather than
appearing literally within the regexp.  That is normally disallowed, since
it is a
potential security risk.  Note that this pragma is ignored when the regular
expression is obtained from tainted data, i.e.  evaluation is always
disallowed with tainted regular expressions.  See \*(L"(?{ code })\*(R" in perlre 
and \*(L"(??{ code })\*(R" in perlre.
.PP
For the purpose of this pragma, interpolation of precompiled regular
expressions (i.e., the result of \f(CW\*(C`qr//\*(C'\fR) is \fInot\fR considered variable
interpolation.  Thus:
.PP
.Vb 1
\&    /foo${pat}bar/
.Ve
.PP
\&\fIis\fR allowed if \f(CW$pat\fR is a precompiled regular expression, even
if \f(CW$pat\fR contains \f(CW\*(C`(?{ ... })\*(C'\fR assertions or \f(CW\*(C`(??{ ... })\*(C'\fR subexpressions.
.SS "'strict' mode"
.IX Subsection "'strict' mode"
Note that this is an experimental feature which may be changed or removed in a
future Perl release.
.PP
When \f(CW\*(C`use re \*(Aqstrict\*(Aq\*(C'\fR is in effect, stricter checks are applied than
otherwise when compiling regular expressions patterns.  These may cause more
warnings to be raised than otherwise, and more things to be fatal instead of
just warnings.  The purpose of this is to find and report at compile time some
things, which may be legal, but have a reasonable possibility of not being the
programmer's actual intent.  This automatically turns on the \f(CW"regexp"\fR
warnings category (if not already on) within its scope.
.PP
As an example of something that is caught under \f(CW\*(C`"strict\*(Aq\*(C'\fR, but not
otherwise, is the pattern
.PP
.Vb 1
\& qr/\exABC/
.Ve
.PP
The \f(CW"\ex"\fR construct without curly braces should be followed by exactly two
hex digits; this one is followed by three.  This currently evaluates as
equivalent to
.PP
.Vb 1
\& qr/\ex{AB}C/
.Ve
.PP
that is, the character whose code point value is \f(CW0xAB\fR, followed by the
letter \f(CW\*(C`C\*(C'\fR.  But since \f(CW\*(C`C\*(C'\fR is a hex digit, there is a reasonable chance
that the intent was
.PP
.Vb 1
\& qr/\ex{ABC}/
.Ve
.PP
that is the single character at \f(CW0xABC\fR.  Under \f(CW\*(Aqstrict\*(Aq\fR it is an error to
not follow \f(CW\*(C`\ex\*(C'\fR with exactly two hex digits.  When not under \f(CW\*(Aqstrict\*(Aq\fR a
warning is generated if there is only one hex digit, and no warning is raised
if there are more than two.
.PP
It is expected that what exactly \f(CW\*(Aqstrict\*(Aq\fR does will evolve over time as we
gain experience with it.  This means that programs that compile under it in
today's Perl may not compile, or may have more or fewer warnings, in future
Perls.  There is no backwards compatibility promises with regards to it.  Also
there are already proposals for an alternate syntax for enabling it.  For
these reasons, using it will raise a \f(CW\*(C`experimental::re_strict\*(C'\fR class warning,
unless that category is turned off.
.PP
Note that if a pattern compiled within \f(CW\*(Aqstrict\*(Aq\fR is recompiled, say by
interpolating into another pattern, outside of \f(CW\*(Aqstrict\*(Aq\fR, it is not checked
again for strictness.  This is because if it works under strict it must work
under non-strict.
.SS "'/flags' mode"
.IX Subsection "'/flags' mode"
When \f(CW\*(C`use re \*(Aq/\f(CIflags\f(CW\*(Aq\*(C'\fR is specified, the given \fIflags\fR are automatically
added to every regular expression till the end of the lexical scope.
\&\fIflags\fR can be any combination of
\&\f(CW\*(Aqa\*(Aq\fR,
\&\f(CW\*(Aqaa\*(Aq\fR,
\&\f(CW\*(Aqd\*(Aq\fR,
\&\f(CW\*(Aqi\*(Aq\fR,
\&\f(CW\*(Aql\*(Aq\fR,
\&\f(CW\*(Aqm\*(Aq\fR,
\&\f(CW\*(Aqn\*(Aq\fR,
\&\f(CW\*(Aqp\*(Aq\fR,
\&\f(CW\*(Aqs\*(Aq\fR,
\&\f(CW\*(Aqu\*(Aq\fR,
\&\f(CW\*(Aqx\*(Aq\fR,
and/or
\&\f(CW\*(Aqxx\*(Aq\fR.
.PP
\&\f(CW\*(C`no re \*(Aq/\f(CIflags\f(CW\*(Aq\*(C'\fR will turn off the effect of \f(CW\*(C`use re \*(Aq/\f(CIflags\f(CW\*(Aq\*(C'\fR for the
given flags.
.PP
For example, if you want all your regular expressions to have /msxx on by
default, simply put
.PP
.Vb 1
\&    use re \*(Aq/msxx\*(Aq;
.Ve
.PP
at the top of your code.
.PP
The character set \f(CW\*(C`/adul\*(C'\fR flags cancel each other out. So, in this example,
.PP
.Vb 4
\&    use re "/u";
\&    "ss" =~ /\exdf/;
\&    use re "/d";
\&    "ss" =~ /\exdf/;
.Ve
.PP
the second \f(CW\*(C`use re\*(C'\fR does an implicit \f(CW\*(C`no re \*(Aq/u\*(Aq\*(C'\fR.
.PP
Similarly,
.PP
.Vb 4
\&    use re "/xx";   # Doubled\-x
\&    ...
\&    use re "/x";    # Single x from here on
\&    ...
.Ve
.PP
Turning on one of the character set flags with \f(CW\*(C`use re\*(C'\fR takes precedence over the
\&\f(CW\*(C`locale\*(C'\fR pragma and the 'unicode_strings' \f(CW\*(C`feature\*(C'\fR, for regular
expressions. Turning off one of these flags when it is active reverts to
the behaviour specified by whatever other pragmata are in scope. For
example:
.PP
.Vb 4
\&    use feature "unicode_strings";
\&    no re "/u"; # does nothing
\&    use re "/l";
\&    no re "/l"; # reverts to unicode_strings behaviour
.Ve
.SS "'debug' mode"
.IX Subsection "'debug' mode"
When \f(CW\*(C`use re \*(Aqdebug\*(Aq\*(C'\fR is in effect, perl emits debugging messages when
compiling and using regular expressions.  The output is the same as that
obtained by running a \f(CW\*(C`\-DDEBUGGING\*(C'\fR\-enabled perl interpreter with the
\&\fB\-Dr\fR switch. It may be quite voluminous depending on the complexity
of the match.  Using \f(CW\*(C`debugcolor\*(C'\fR instead of \f(CW\*(C`debug\*(C'\fR enables a
form of output that can be used to get a colorful display on terminals
that understand termcap color sequences.  Set \f(CW$ENV{PERL_RE_TC}\fR to a
comma-separated list of \f(CW\*(C`termcap\*(C'\fR properties to use for highlighting
strings on/off, pre-point part on/off.
See \*(L"Debugging Regular Expressions\*(R" in perldebug for additional info.
.PP
\&\fB\s-1NOTE\s0\fR that the exact format of the \f(CW\*(C`debug\*(C'\fR mode is \fB\s-1NOT\s0\fR considered
to be an officially supported \s-1API\s0 of Perl. It is intended for debugging
only and may change as the core development team deems appropriate
without notice or deprecation in any release of Perl, major or minor.
Any documentation of the output is purely advisory.
.PP
As of 5.9.5 the directive \f(CW\*(C`use re \*(Aqdebug\*(Aq\*(C'\fR and its equivalents are
lexically scoped, as the other directives are.  However they have both
compile-time and run-time effects.
.PP
See \*(L"Pragmatic Modules\*(R" in perlmodlib.
.SS "'Debug' mode"
.IX Subsection "'Debug' mode"
Similarly \f(CW\*(C`use re \*(AqDebug\*(Aq\*(C'\fR produces debugging output, the difference
being that it allows the fine tuning of what debugging output will be
emitted. Options are divided into three groups, those related to
compilation, those related to execution and those related to special
purposes.
.PP

perl-5.36.0/ext/re/re.c
/*
 * This file was generated automatically by ExtUtils::ParseXS version 3.45 from the
 * contents of re.xs. Do not edit this file, edit re.xs instead.
 *
 *    ANY CHANGES MADE HERE WILL BE LOST!
 *
 */

#line 1 "re.xs"
#if defined(PERL_EXT_RE_DEBUG) && !defined(DEBUGGING)
#  define DEBUGGING
#  define DEBUGGING_RE_ONLY
#endif

#define PERL_NO_GET_CONTEXT
#include "EXTERN.h"
#include "perl.h"
#include "XSUB.h"
#include "re_comp.h"

#undef dXSBOOTARGSXSAPIVERCHK
/* skip API version checking due to different interp struct size but,
   this hack is until GitHub issue #14169 is resolved */
#define dXSBOOTARGSXSAPIVERCHK dXSBOOTARGSNOVERCHK

START_EXTERN_C

extern REGEXP*	my_re_compile (pTHX_ SV * const pattern, const U32 pm_flags);
extern REGEXP*	my_re_op_compile (pTHX_ SV ** const patternp, int pat_count,
		    OP *expr, const regexp_engine* eng, REGEXP *volatile old_re,
		     bool *is_bare_re, U32 rx_flags, U32 pm_flags);

extern I32	my_regexec (pTHX_ REGEXP * const prog, char* stringarg, char* strend,
			    char* strbeg, SSize_t minend, SV* screamer,
			    void* data, U32 flags);

extern char*	my_re_intuit_start(pTHX_
                    REGEXP * const rx,
                    SV *sv,
                    const char * const strbeg,
                    char *strpos,
                    char *strend,
                    const U32 flags,
                    re_scream_pos_data *data);

extern SV*	my_re_intuit_string (pTHX_ REGEXP * const prog);

extern void	my_regfree (pTHX_ REGEXP * const r);

extern void	my_reg_numbered_buff_fetch(pTHX_ REGEXP * const rx, const I32 paren,
					   SV * const usesv);
extern void	my_reg_numbered_buff_store(pTHX_ REGEXP * const rx, const I32 paren,
					   SV const * const value);
extern I32	my_reg_numbered_buff_length(pTHX_ REGEXP * const rx,
					    const SV * const sv, const I32 paren);

extern SV*	my_reg_named_buff(pTHX_ REGEXP * const, SV * const, SV * const,
                              const U32);
extern SV*	my_reg_named_buff_iter(pTHX_ REGEXP * const rx,
                                   const SV * const lastkey, const U32 flags);

extern SV*      my_reg_qr_package(pTHX_ REGEXP * const rx);
#if defined(USE_ITHREADS)
extern void*	my_regdupe (pTHX_ REGEXP * const r, CLONE_PARAMS *param);
#endif
extern void     my_regprop(pTHX_
    const regexp *prog, SV* sv, const regnode* o,
    const regmatch_info *reginfo, const RExC_state_t *pRExC_state
);

EXTERN_C const struct regexp_engine my_reg_engine;
EXTERN_C const struct regexp_engine wild_reg_engine;

END_EXTERN_C

const struct regexp_engine my_reg_engine = { 
        my_re_compile, 
        my_regexec, 
        my_re_intuit_start, 
        my_re_intuit_string, 
        my_regfree, 
        my_reg_numbered_buff_fetch,
        my_reg_numbered_buff_store,
        my_reg_numbered_buff_length,
        my_reg_named_buff,
        my_reg_named_buff_iter,
        my_reg_qr_package,
#if defined(USE_ITHREADS)
        my_regdupe,
#endif
        my_re_op_compile,
};

/* For use with Unicode property wildcards, when we want to see the compilation
 * of the wildcard subpattern, but don't want to see the matching process.  All
 * but the compilation are the regcomp.c/regexec.c functions which aren't
 * subject to 'use re' */
const struct regexp_engine wild_reg_engine = {
        my_re_compile,
        Perl_regexec_flags,
        Perl_re_intuit_start,
        Perl_re_intuit_string,
        Perl_regfree_internal,
        Perl_reg_numbered_buff_fetch,
        Perl_reg_numbered_buff_store,
        Perl_reg_numbered_buff_length,
        Perl_reg_named_buff,
        Perl_reg_named_buff_iter,
        Perl_reg_qr_package,
#if defined(USE_ITHREADS)
        Perl_regdupe_internal,
#endif
        my_re_op_compile,
};

#define newSVbool_(x) newSViv((x) ? 1 : 0)

#line 119 "re.c"
#ifndef PERL_UNUSED_VAR
#  define PERL_UNUSED_VAR(var) if (0) var = var
#endif

#ifndef dVAR
#  define dVAR		dNOOP
#endif


/* This stuff is not part of the API! You have been warned. */
#ifndef PERL_VERSION_DECIMAL
#  define PERL_VERSION_DECIMAL(r,v,s) (r*1000000 + v*1000 + s)
#endif
#ifndef PERL_DECIMAL_VERSION
#  define PERL_DECIMAL_VERSION \
	  PERL_VERSION_DECIMAL(PERL_REVISION,PERL_VERSION,PERL_SUBVERSION)
#endif
#ifndef PERL_VERSION_GE
#  define PERL_VERSION_GE(r,v,s) \
	  (PERL_DECIMAL_VERSION >= PERL_VERSION_DECIMAL(r,v,s))
#endif
#ifndef PERL_VERSION_LE
#  define PERL_VERSION_LE(r,v,s) \
	  (PERL_DECIMAL_VERSION <= PERL_VERSION_DECIMAL(r,v,s))
#endif

/* XS_INTERNAL is the explicit static-linkage variant of the default
 * XS macro.
 *
 * XS_EXTERNAL is the same as XS_INTERNAL except it does not include
 * "STATIC", ie. it exports XSUB symbols. You probably don't want that
 * for anything but the BOOT XSUB.
 *
 * See XSUB.h in core!
 */


/* TODO: This might be compatible further back than 5.10.0. */
#if PERL_VERSION_GE(5, 10, 0) && PERL_VERSION_LE(5, 15, 1)
#  undef XS_EXTERNAL
#  undef XS_INTERNAL
#  if defined(__CYGWIN__) && defined(USE_DYNAMIC_LOADING)
#    define XS_EXTERNAL(name) __declspec(dllexport) XSPROTO(name)
#    define XS_INTERNAL(name) STATIC XSPROTO(name)
#  endif
#  if defined(__SYMBIAN32__)
#    define XS_EXTERNAL(name) EXPORT_C XSPROTO(name)
#    define XS_INTERNAL(name) EXPORT_C STATIC XSPROTO(name)
#  endif
#  ifndef XS_EXTERNAL
#    if defined(HASATTRIBUTE_UNUSED) && !defined(__cplusplus)
#      define XS_EXTERNAL(name) void name(pTHX_ CV* cv __attribute__unused__)
#      define XS_INTERNAL(name) STATIC void name(pTHX_ CV* cv __attribute__unused__)
#    else
#      ifdef __cplusplus
#        define XS_EXTERNAL(name) extern "C" XSPROTO(name)
#        define XS_INTERNAL(name) static XSPROTO(name)
#      else
#        define XS_EXTERNAL(name) XSPROTO(name)
#        define XS_INTERNAL(name) STATIC XSPROTO(name)
#      endif
#    endif
#  endif
#endif

/* perl >= 5.10.0 && perl <= 5.15.1 */


/* The XS_EXTERNAL macro is used for functions that must not be static
 * like the boot XSUB of a module. If perl didn't have an XS_EXTERNAL
 * macro defined, the best we can do is assume XS is the same.
 * Dito for XS_INTERNAL.
 */
#ifndef XS_EXTERNAL
#  define XS_EXTERNAL(name) XS(name)
#endif
#ifndef XS_INTERNAL
#  define XS_INTERNAL(name) XS(name)
#endif

/* Now, finally, after all this mess, we want an ExtUtils::ParseXS
 * internal macro that we're free to redefine for varying linkage due
 * to the EXPORT_XSUB_SYMBOLS XS keyword. This is internal, use
 * XS_EXTERNAL(name) or XS_INTERNAL(name) in your code if you need to!
 */

#undef XS_EUPXS
#if defined(PERL_EUPXS_ALWAYS_EXPORT)
#  define XS_EUPXS(name) XS_EXTERNAL(name)
#else
   /* default to internal */
#  define XS_EUPXS(name) XS_INTERNAL(name)
#endif

#ifndef PERL_ARGS_ASSERT_CROAK_XS_USAGE
#define PERL_ARGS_ASSERT_CROAK_XS_USAGE assert(cv); assert(params)

/* prototype to pass -Wmissing-prototypes */
STATIC void
S_croak_xs_usage(const CV *const cv, const char *const params);

STATIC void
S_croak_xs_usage(const CV *const cv, const char *const params)
{
    const GV *const gv = CvGV(cv);

    PERL_ARGS_ASSERT_CROAK_XS_USAGE;

    if (gv) {
        const char *const gvname = GvNAME(gv);
        const HV *const stash = GvSTASH(gv);
        const char *const hvname = stash ? HvNAME(stash) : NULL;

        if (hvname)
	    Perl_croak_nocontext("Usage: %s::%s(%s)", hvname, gvname, params);
        else
	    Perl_croak_nocontext("Usage: %s(%s)", gvname, params);
    } else {
        /* Pants. I don't think that it should be possible to get here. */
	Perl_croak_nocontext("Usage: CODE(0x%" UVxf ")(%s)", PTR2UV(cv), params);
    }
}
#undef  PERL_ARGS_ASSERT_CROAK_XS_USAGE

#define croak_xs_usage        S_croak_xs_usage

#endif

/* NOTE: the prototype of newXSproto() is different in versions of perls,
 * so we define a portable version of newXSproto()
 */
#ifdef newXS_flags
#define newXSproto_portable(name, c_impl, file, proto) newXS_flags(name, c_impl, file, proto, 0)
#else
#define newXSproto_portable(name, c_impl, file, proto) (PL_Sv=(SV*)newXS(name, c_impl, file), sv_setpv(PL_Sv, proto), (CV*)PL_Sv)
#endif /* !defined(newXS_flags) */

#if PERL_VERSION_LE(5, 21, 5)
#  define newXS_deffile(a,b) Perl_newXS(aTHX_ a,b,file)
#else
#  define newXS_deffile(a,b) Perl_newXS_deffile(aTHX_ a,b)
#endif

#line 263 "re.c"

XS_EUPXS(XS_re_install); /* prototype to pass -Wmissing-prototypes */
XS_EUPXS(XS_re_install)
{
    dVAR; dXSARGS;
    if (items != 0)
       croak_xs_usage(cv,  "");
    PERL_UNUSED_VAR(ax); /* -Wall */
    SP -= items;
    {
#line 114 "re.xs"
        PL_colorset = 0;	/* Allow reinspection of ENV. */
        /* PL_debug |= DEBUG_r_FLAG; */
	XPUSHs(sv_2mortal(newSViv(PTR2IV(&my_reg_engine))));
#line 278 "re.c"
	PUTBACK;
	return;
    }
}


XS_EUPXS(XS_re_regmust); /* prototype to pass -Wmissing-prototypes */
XS_EUPXS(XS_re_regmust)
{
    dVAR; dXSARGS;
    if (items != 1)
       croak_xs_usage(cv,  "sv");
    PERL_UNUSED_VAR(ax); /* -Wall */
    SP -= items;
    {
	SV *	sv = ST(0)
;
#line 123 "re.xs"
    REGEXP *re;
#line 298 "re.c"
#line 125 "re.xs"
{
    if ((re = SvRX(sv)) /* assign deliberate */
       /* only for re engines we know about */
       && (   RX_ENGINE(re) == &my_reg_engine
           || RX_ENGINE(re) == &wild_reg_engine
           || RX_ENGINE(re) == &PL_core_reg_engine))
    {
        SV *an = &PL_sv_no;
        SV *fl = &PL_sv_no;
        if (RX_ANCHORED_SUBSTR(re)) {
            an = sv_2mortal(newSVsv(RX_ANCHORED_SUBSTR(re)));
        } else if (RX_ANCHORED_UTF8(re)) {
            an = sv_2mortal(newSVsv(RX_ANCHORED_UTF8(re)));
        }
        if (RX_FLOAT_SUBSTR(re)) {
            fl = sv_2mortal(newSVsv(RX_FLOAT_SUBSTR(re)));
        } else if (RX_FLOAT_UTF8(re)) {
            fl = sv_2mortal(newSVsv(RX_FLOAT_UTF8(re)));
        }
        EXTEND(SP, 2);
        PUSHs(an);
        PUSHs(fl);
        XSRETURN(2);
    }
    XSRETURN_UNDEF;
}
#line 326 "re.c"
	PUTBACK;
	return;
    }
}


XS_EUPXS(XS_re_optimization); /* prototype to pass -Wmissing-prototypes */
XS_EUPXS(XS_re_optimization)
{
    dVAR; dXSARGS;
    if (items != 1)
       croak_xs_usage(cv,  "sv");
    {
	SV *	sv = ST(0)
;
#line 157 "re.xs"
    REGEXP *re;
    regexp *r;
    struct reg_substr_datum * data;
    HV *hv;
#line 347 "re.c"
	SV *	RETVAL;
#line 162 "re.xs"
{
    re = SvRX(sv);
    if (!re) {
        XSRETURN_UNDEF;
    }

    /* only for re engines we know about */
    if (   RX_ENGINE(re) != &my_reg_engine
        && RX_ENGINE(re) != &wild_reg_engine
        && RX_ENGINE(re) != &PL_core_reg_engine)
    {
        XSRETURN_UNDEF;
    }

    if (!PL_colorset) {
        reginitcolors();
    }

    r = ReANY(re);
    hv = newHV();

    hv_stores(hv, "minlen", newSViv(r->minlen));
    hv_stores(hv, "minlenret", newSViv(r->minlenret));
    hv_stores(hv, "gofs", newSViv(r->gofs));

    data = &r->substrs->data[0];
    hv_stores(hv, "anchored", data->substr
            ? newSVsv(data->substr) : &PL_sv_undef);
    hv_stores(hv, "anchored utf8", data->utf8_substr
            ? newSVsv(data->utf8_substr) : &PL_sv_undef);
    hv_stores(hv, "anchored min offset", newSViv(data->min_offset));
    hv_stores(hv, "anchored max offset", newSViv(data->max_offset));
    hv_stores(hv, "anchored end shift", newSViv(data->end_shift));

    data = &r->substrs->data[1];
    hv_stores(hv, "floating", data->substr
            ? newSVsv(data->substr) : &PL_sv_undef);
    hv_stores(hv, "floating utf8", data->utf8_substr
            ? newSVsv(data->utf8_substr) : &PL_sv_undef);
    hv_stores(hv, "floating min offset", newSViv(data->min_offset));
    hv_stores(hv, "floating max offset", newSViv(data->max_offset));
    hv_stores(hv, "floating end shift", newSViv(data->end_shift));

    hv_stores(hv, "checking", newSVpv(
        (!r->check_substr && !r->check_utf8)
            ? "none"
        : (    r->check_substr == r->substrs->data[1].substr
            && r->check_utf8   == r->substrs->data[1].utf8_substr
        )
            ? "floating"
        : "anchored"
    , 0));

    hv_stores(hv, "noscan", newSVbool_(r->intflags & PREGf_NOSCAN));
    hv_stores(hv, "isall", newSVbool_(r->extflags & RXf_CHECK_ALL));
    hv_stores(hv, "anchor SBOL", newSVbool_(r->intflags & PREGf_ANCH_SBOL));
    hv_stores(hv, "anchor MBOL", newSVbool_(r->intflags & PREGf_ANCH_MBOL));
    hv_stores(hv, "anchor GPOS", newSVbool_(r->intflags & PREGf_ANCH_GPOS));
    hv_stores(hv, "skip", newSVbool_(r->intflags & PREGf_SKIP));
    hv_stores(hv, "implicit", newSVbool_(r->intflags & PREGf_IMPLICIT));

    {
        RXi_GET_DECL(r, ri);
        if (ri->regstclass) {
            SV* sv = newSV(0);
            /* not Perl_regprop, we must have the DEBUGGING version */
            my_regprop(aTHX_ r, sv, ri->regstclass, NULL, NULL);
            hv_stores(hv, "stclass", sv);
        } else {
            hv_stores(hv, "stclass", &PL_sv_undef);
        }
    }

    RETVAL = newRV_noinc((SV *)hv);
}
#line 425 "re.c"
	RETVAL = sv_2mortal(RETVAL);
	ST(0) = RETVAL;
    }
    XSRETURN(1);
}

#ifdef __cplusplus
extern "C"
#endif
XS_EXTERNAL(boot_re); /* prototype to pass -Wmissing-prototypes */
XS_EXTERNAL(boot_re)
{
#if PERL_VERSION_LE(5, 21, 5)
    dVAR; dXSARGS;
#else
    dVAR; dXSBOOTARGSXSAPIVERCHK;
#endif
#if PERL_VERSION_LE(5, 8, 999) /* PERL_VERSION_LT is 5.33+ */
    char* file = __FILE__;
#else
    const char* file = __FILE__;
#endif

    PERL_UNUSED_VAR(file);

    PERL_UNUSED_VAR(cv); /* -W */
    PERL_UNUSED_VAR(items); /* -W */
#if PERL_VERSION_LE(5, 21, 5)
    XS_VERSION_BOOTCHECK;
#  ifdef XS_APIVERSION_BOOTCHECK
    XS_APIVERSION_BOOTCHECK;
#  endif
#endif

        newXS_deffile("re::install", XS_re_install);
        (void)newXSproto_portable("re::regmust", XS_re_regmust, file, "$");
        (void)newXSproto_portable("re::optimization", XS_re_optimization, file, "$");
#if PERL_VERSION_LE(5, 21, 5)
#  if PERL_VERSION_GE(5, 9, 0)
    if (PL_unitcheckav)
        call_list(PL_scopestack_ix, PL_unitcheckav);
#  endif
    XSRETURN_YES;
#else
    Perl_xs_boot_epilog(aTHX_ ax);
#endif
}

